### PR TITLE
[SR-5869][Parse/Sema] Expand recovery support for trailing closure in statement conditions

### PR DESCRIPTION
**Pull request by rintaro**
Fixes: [SR-5869](https://bugs.swift.org/browse/SR-5869)

Formerly, the recovering logic only handled trailing closures at outermost position.
This PR expands it to handle trailing closures in middle of expressions:
```
test.swift:2:16: error: trailing closure requires parentheses for disambiguation in this context
if nums.filter { $0 % 2 == 0 }.isEmpty { 
              ~^
              (               )
test.swift:5:16: error: trailing closure requires parentheses for disambiguation in this context
if nums.filter { $0 % 2 == 0 } == [2, 4] { 
              ~^
              (               )
```

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Sema/MiscDiagnostics.cpp [hunk 1]
```diff
@@ -2871,13 +2871,31 @@ static void checkStmtConditionTrailingClosure(TypeChecker &TC, const Expr *E) {
...
+      switch (E->getKind()) {
+      case ExprKind::Closure:
+      case ExprKind::Paren:
+      case ExprKind::Tuple:
+      case ExprKind::TupleShuffle:
+      case ExprKind::ObjectLiteral:
+      case ExprKind::Array:
+      case ExprKind::Dictionary:
+        // Don't dig into expression which syntactically enclose other
+        // expressinos. e.g. [...], (...), and { ... }
```
Comment (harlanhaskins) [inline lib/Sema/MiscDiagnostics.cpp]: Typo: `expressinos`

Comment (natecook1000) [reply lib/Sema/MiscDiagnostics.cpp]: I could really go for an expressino right now.

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by rintaro** [comment]
Closing. This has been superseded by  https://github.com/apple/swift/pull/25165 which is reverted.
We should make this diagnostic warning (https://github.com/apple/swift/pull/25165#issuecomment-506898762).

**Comment by owenv** [comment]
I’d forgotten about #25165, thanks for the reminder! I should be able to get back to that pretty soon and turn it into a warning

**Pull request closed without merging**